spec/formatting.md

```diff
@@ -225,40 +225,41 @@ the following steps are taken:
    A _local-declaration_ binds the resolved value of an _expression_
    to a _variable_.
    Thus, the output of one _function_ is potentially the _operand_
-   of another _function_. In other words, formatting functions
-   compose with each other.
+   of another _function_,
+   or the value of one of the _options_ for another function.
    For example, in
    ```
    .input {$n :number minIntegerDigits=3}
    .local $n1 = {$n :number maxFractionDigits=3}
    ```
-   the second call to `:number` composes with the first call.
+   the output of the first call to `:number`
+   is the input of the second call to `:number`.
 
-   In addition, selector functions compose with formatting functions
-   in the sense that a selector function's _operand_
-   may be the output of any formatting function.
-
-   Implementations SHOULD provide a means for formatting functions
-   to compose with each other
-   and for formatting functions to compose with selector functions.
    Implementations that provide a means for defining custom functions
-   SHOULD provide a means for those functions to return values
-   that contain enough information
+   SHOULD provide a means for function implementations
+   to return values that contain enough information
    (e.g. the resolved _operand_ and _option_ values
    that the function was called with)
-   to be used as inputs to subsequent function calls.
-   For example, an implementation in a typed programming language
-   MAY define an interface that custom functions implement.
+   to be used as arguments to subsequent calls
+   to the function implementations.
+   For example, a MessageFormat implementation in a typed programming language
+   MAY define an interface for custom function implementations.
    Such an interface SHOULD define an implementation-specific
-   argument type `T` and return type `U` for custom formatting functions
+   argument type `T` and return type `U`
+   for implementations of formatting functions
    such that `U` can be coerced to `T` without loss of information.
```

Comment:
Look at the following mutating function. Do we expect to be able to extract

.local $foo {$date :extractDay calendar=georgian}
.local $fii {$foo :extractMonth calendar=georgian}

Would the above text mean that  :extractDay SHOULD have a "return type" such that it preserves the month value from $date?

Reply:
No; to be concrete, I'll use a simplified version of the C++ implementation.

`:extractDay` and `:extractMonths` would be implemented as classes that provide a `format()` method. The type of that method (ignoring options and context, for simplicity) is:

```
FormattedPlaceholder format(FormattedPlaceholder&& argument);
```

A particular instance of the type `FormattedPlaceholder` could contain any set of options, or none.

Sorry for talking in implementation terms, but I'm not sure how else to say it, given that my goal here is to suggest that implementors not do the "obvious" thing, which would be to define something like a `FormatterInput` and `FormatterOutput` type, which are not coercible to each other, and define the interface like:

```
FormatterOutput format(FormatterInput&& argument);
```

(I would expect every implementation to have some sort of interface between the formatter and calls to custom functions (I say "custom" because built-in functions could be handled inside the formatter), so I think it's meaningful to refer to it in this note. In a unityped implementation language like JavaScript, there's less of a hazard since there's only one type, trivially guaranteeing the property stated in the note.)

Reply:
This is still in an example and must not be normative. Even if it weren't an example, it's way too specific to put in the specification.

If we want to provide guidance to users/implementers, instead of pouring stuff into the spec, we should write some user guide material.